pydantic/json_schema.py

```diff
@@ -995,7 +981,10 @@ def function_before_schema(self, schema: core_schema.BeforeValidatorFunctionSche
         Returns:
             The generated JSON schema.
         """
-        return self._function_schema(schema)
+        if self._mode == 'validation' and (input_schema := schema.get('metadata', {}).get('input_schema')):
+            return self.generate_inner(input_schema)
+
+        return self.generate_inner(schema['schema'])
```

Comment:
We'll want to use the `CoreMetadataHandler` thing here once we make the switch to support `json_schema_input_type` there -- perhaps we also want to name the attribute with `json` in the name for consistency + clarity